### DISCUSSION TITLE
Distributed+Mergetree Cluster architure , ask for advice on data sending knowledge

### DISCUSSION THREAD

**Discussion by shelly-ctrl** [ClickHouse/ClickHouse#59843] [Q&A] [answered]
I have a clickhouse six-shard two-replicas cluster, two replicas of which are due to disk hardware failure, the local table uses mergetree, the distributed table to do query insert operations, now I insert data into a node, the data will also send data to the failed replicas, I want to know that the failed node can not receive data, Will the unreceived data be cached in the node where the data is inserted, and if so, where is the data stored? After the bad node is repaired, will the data not received during this period still be sent to the node?
thanks!

**Comment by mlex** [comment, marked as answer]
The behaviour depends on the setting `distributed_foreground_insert` (see [docs](https://clickhouse.com/docs/en/operations/settings/settings#distributed_foreground_insert)).

By default (distributed_foreground_insert=0), the clickhouse node that receives the INSERT will write the data to the local filesystem. See https://clickhouse.com/docs/en/engines/table-engines/special/distributed#distributed-writing-data for how this works. The node will then send the data asynchronously (in a background thread) to the targe shard. I'm not sure if the background distributed send background thread will try out another replica if one replica returned a failure. I assume it does. 

So to answer your questions:

> Will the unreceived data be cached in the node where the data is inserted, and if so, where is the data stored? 

Yes. By default on the disk in `/var/lib/clickhouse/data/database/table/`. You can monitor it via the `system.distribution_queue` system table (see [docs](https://clickhouse.com/docs/en/operations/system-tables/distribution_queue)). 

> After the bad node is repaired, will the data not received during this period still be sent to the node?

Yes
